common/app/common/Logback/KinesisAdapter.scala

```diff
@@ -0,0 +1,49 @@
+                        package common.Logback
+
+import java.util.concurrent.ThreadPoolExecutor
+
+import akka.actor.ActorSystem
+import akka.dispatch.MessageDispatcher
+import akka.pattern.CircuitBreaker
+import ch.qos.logback.classic.spi.ILoggingEvent
+import com.amazonaws.ClientConfiguration
+import com.amazonaws.auth.AWSCredentialsProvider
+import com.amazonaws.retry.{PredefinedRetryPolicies, RetryPolicy}
+import com.amazonaws.services.kinesis.AmazonKinesisAsyncClient
+import com.gu.logback.appender.kinesis.KinesisAppender
+import scala.concurrent.duration._
+import scala.concurrent.Future
+
+// LogbackOperationsPool must be wired as a singleton
+class LogbackOperationsPool(val actorSystem: ActorSystem)  {
+  val logbackOperations: MessageDispatcher = actorSystem.dispatchers.lookup("akka.logback-operations")
+}
+
+class SafeBlockingKinesisAppender(logbackOperations: LogbackOperationsPool) extends KinesisAppender[ILoggingEvent] {
+
+  private val breaker = new CircuitBreaker(
```

Comment:
I'd prefer to use a Semaphore to prevent the number concurrent operations from leaking @nicl 
The circuit breaker helps, but doesn't actually prevent blocking concurrent operations from leaking and filling up the threadpool.

Reply:
I agree, the CB only kicks in when you have a problem - in this sense a semaphore is better. But it should still act as a bulkhead - if the threadpool is saturated the CB should kick in until it becomes responsive again. The 'semaphore' in this case is (loosely) the number of threads in the EC.

I like CBs because the semantics are well understood, and Akka has one pre-built which is well tested and used widely already in Grauniad land.